stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -525,17 +525,35 @@ end
 
 #Eigensystem
 eigvals(D::Diagonal{<:Number}; permute::Bool=true, scale::Bool=true) = D.diag
-eigvals(D::Diagonal; permute::Bool=true, scale::Bool=true) =
-    [eigvals(x) for x in D.diag] #For block matrices, etc.
+eigvals(D::Diagonal; permute::Bool=true, scale::Bool=true) = vcat([eigvals(d; permute=permute, scale=scale) for d in D.diag]...)
 eigvecs(D::Diagonal{<:Number}) = Matrix{eltype(D)}(I, size(D))
-eigvecs(D::Diagonal) = [eigvecs(x) for x in D.diag]
+function eigvecs(D::Diagonal)
+    vec = vcat([vcat([zeros(size(x,2)) for x in D.diag]) for i=1:length(eigvals(D))])
+    ind = 0
+    for i=1:length(D.diag)
+        s = size(D.diag[i],2)
+        x = eigvecs(D.diag[i])
+        indices = findall(m -> m==D.diag[i], D)
+
+        for j in indices
```

Comment:
I'm afraid I don't quite understand this loop and the `findall` above. I thought you simply go through the diagonal elements (the `i`-loop above), and in this loop, you go through the eigenvectors of `D.diag[i]` (the `k`-loop below).